src/libostree/ostree-repo-libarchive.c

```diff
@@ -45,269 +47,715 @@ propagate_libarchive_error (GError      **error,
                "%s", archive_error_string (a));
 }
 
+static inline const char *
+path_relative (const char *src)
+{
+  /* One issue here is that some archives almost record the pathname as just a
+   * string and don't need to actually encode parent/child relationships in the
+   * archive. For us however, this will be important. So we do our best to deal
+   * with non-conventional paths (but note that we don't try to be sophisticated
+   * enough to handle e.g. ../ & ./ in the middle of normal dir comps, which
+   * would probably require a stack). Also important, we relativize the path. */
```

Comment:
Note `ot_util_path_split_validate()` *does* intentionally explicitly error out if we find `..` as a path component.  See also `ot_util_filename_validate()`.

I think with the current logic it'd be possible to have `foo/../../../../bar`...and we wouldn't trip any assertions =/

We really need unit tests for this function.

I submitted https://github.com/ostreedev/ostree/pull/283